### DISCUSSION TITLE
rename a metric tag, how ?

### DISCUSSION THREAD

**Discussion by StefanSa** [vectordotdev/vector#12067] [Q&A] [answered]
Hi there,
i get these metrics from the telegraf client.
```
{
  "fields": {
    "usage_guest": 0,
    "usage_guest_nice": 0,
    "usage_idle": 99.82499999983702,
    "usage_iowait": 0,
    "usage_irq": 0,
    "usage_nice": 0,
    "usage_softirq": 0,
    "usage_steal": 0,
    "usage_system": 0.024999999999977263,
    "usage_user": 0.14999999999986358
  },
  "name": "cpu",
  "tags": {
    "cpu": "cpu-total",
    "host": "os-dash01"
  },
  "timestamp": 1649056860
}
```
The array name `fields` would have to be renamed with the `contents` of the field `name` (cpu). What is the best way to do this ?

That would be the goal:
```
{
  "cpu": {
    "usage_guest": 0,
    "usage_guest_nice": 0,
    "usage_idle": 99.82499999983702,
    "usage_iowait": 0,
    "usage_irq": 0,
    "usage_nice": 0,
    "usage_softirq": 0,
    "usage_steal": 0,
    "usage_system": 0.024999999999977263,
    "usage_user": 0.14999999999986358
  },
  "name": "cpu",
  "tags": {
    "cpu": "cpu-total",
    "host": "os-dash01"
  },
  "timestamp": 1649056860
}
```

**Comment by jszwedko** [comment]
Hi @StefanSa ,

You can use `remap` with a VRL program like:

```
.cpu = del(.fields)
```

**Comment by StefanSa** [comment, marked as answer]
Hi @jszwedko , good evening from Germany.
Thank you for your quick reply. I already know this solution / example.
The field `name` is dynamic and can have different values like `cpu, system, ram` etc.
These different values of `name` should overwrite the name of `fields`.
How do i do this ?

**Reply by jszwedko** [reply]
Hi @StefanSa ,

Ah, in that case, I think something like:

```
set(., [.name], del(.fields))
```

This makes use of the dynamic path lookup function `set` (https://vector.dev/docs/reference/vrl/functions/#path-functions).

**Reply by StefanSa** [reply]
@jszwedko 
Perfect, I've thought of that too.
But my tests had not been successful. 
Thanks for the competent and quick help, but especially for `vector`.

**Reply by jszwedko** [reply]
No worries! I neglected the error handling as well. This function requires error handling because VRL can't be sure that `.name` is a string or integer (or doesn't exist at all). I guess my question would be: what do you want to have happen if `.name` is not a string or is not set? If you want the `remap` to fail completely, you can just do `set!(., [.name], del(.fields))`. If you want to skip the rename, you could do:

```
name, err = string(.name)
if err != null {
   set(., [name], del(.fields)) # note that name lacks the `.` here since we are referring to the variable now
}
```

This will only do the rename if `.name` is set and is a string (or integer, technically).

**Reply by StefanSa** [reply]
@jszwedko 
`set!(., [.name], del(.fields)). ` 
is unfortunately not the solution.
fields is completely deleted, but not renamed as wanted.

```
{
  "@timestamp": "2022-04-04T19:57:10Z",
  "name": "cpu",
  "tags": {
    "cpu": "cpu-total",
    "host": "os-dash01"
  }
}
```

**Reply by jszwedko** [reply]
Apologies, I also forgot `set` operates by-value. So you'll need to do `. = set!(., [.name], del(.fields))`.

**Reply by StefanSa** [reply]
@jszwedko 
Perfect, now everything OK.
Thanks again !
